<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<HTML>
 <HEAD>
   <TITLE> thread-safety definition
   </TITLE>
   <LINK REL="Index" HREF="index.html" >
   <LINK REL="made" HREF="mailto:c%2B%2B-pthreads%40codesourcery.com?Subject=Re%3A%20thread-safety%20definition&In-Reply-To=%3Cllogzc38.fsf%40boost-consulting.com%3E">
   <META NAME="robots" CONTENT="index,nofollow">
   <style type="text/css">
       pre {
           white-space: pre-wrap;       /* css-2.1, curent FF, Opera, Safari */
           }
   </style>
   <META http-equiv="Content-Type" content="text/html; charset=us-ascii">
   <LINK REL="Previous"  HREF="000250.html">
   <LINK REL="Next"  HREF="000187.html">
 </HEAD>
 <BODY BGCOLOR="#ffffff">
   <H1>thread-safety definition</H1>
    <B>David Abrahams</B> 
    <A HREF="mailto:c%2B%2B-pthreads%40codesourcery.com?Subject=Re%3A%20thread-safety%20definition&In-Reply-To=%3Cllogzc38.fsf%40boost-consulting.com%3E"
       TITLE="thread-safety definition">dave at boost-consulting.com
       </A><BR>
    <I>Sat Jan 10 04:01:15 UTC 2004</I>
    <P><UL>
        <LI>Previous message: <A HREF="000250.html">[c++-pthreads] Re: thread-safety definition
</A></li>
        <LI>Next message: <A HREF="000187.html">[c++-pthreads] Re: thread-safety definition
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#186">[ date ]</a>
              <a href="thread.html#186">[ thread ]</a>
              <a href="subject.html#186">[ subject ]</a>
              <a href="author.html#186">[ author ]</a>
         </LI>
       </UL>
    <HR>  
<!--beginarticle-->
<PRE>Dave Butenhof &lt;<A HREF="http://sourcerytools.com/cgi-bin/mailman/listinfo/c++-pthreads">David.Butenhof at hp.com</A>&gt; writes:

&gt;&gt;&gt;&gt;<i>If you make cancellations behave sufficiently like an ordinary C++
</I>&gt;&gt;&gt;&gt;<i>exception (either of Nathan's or Jason's models would do that I think)
</I>&gt;&gt;&gt;&gt;<i>then it's neither &quot;IMPOSSIBLE&quot; nor even difficult.  A great deal of
</I>&gt;&gt;&gt;&gt;<i>thread-safe exception-safe C++ library code would behave perfectly
</I>&gt;&gt;&gt;&gt;<i>well under those conditions.
</I>&gt;&gt;&gt;&gt;<i>
</I>&gt;&gt;&gt;<i>Well, it should be no surprise that this discussion has fragmented
</I>&gt;&gt;&gt;<i>into too many different little pathways for anyone to keep track of
</I>&gt;&gt;&gt;<i>them.
</I>&gt;&gt;&gt;<i>
</I>&gt;&gt;&gt;<i>If cancel is implemented completely as a &quot;full stature&quot; C++ exception,
</I>&gt;&gt;&gt;<i>and none of the &quot;exception safe&quot; code does anything silly like
</I>&gt;&gt;&gt;<i>&quot;catch(...)&quot; without a re-throw, then, yes, it's likely to be
</I>&gt;&gt;&gt;<i> reasonably cancel-safe.
</I>&gt;&gt;&gt;<i>
</I>&gt;&gt;<i>Even if only *some* of it does that, Nathan and Jason's models are
</I>&gt;&gt;<i>both likely to result in an eventual cancellation.
</I>&gt;&gt;<i>
</I>&gt;<i> I need to start keeping a scorecard. Lessee here, Nathan's model I
</I>&gt;<i> believe could be trivially characterized as &quot;return failure status on
</I>&gt;<i> any cancellation point until you hit a C++ function allowed to
</I>&gt;<i> throw&quot;. This would apparently disable any cleanup operations involving
</I>&gt;<i> cancellation points (they'd fail) at least prior to the first throw
</I>&gt;<i> point -- and it's not precisely clear how the &quot;pending cancel&quot; state
</I>&gt;<i> is to be managed even beyond that point. While it might eventually
</I>&gt;<i> lead to cancellation, it won't reliably perform cleanup. 
</I>
I'm having trouble understanding.  How does a &quot;pending cancel state&quot;
perform cleanup?  And what do &quot;reliably&quot; and &quot;cleanup&quot; mean to you?

&gt;<i> This seems to me to be more complicated, less predictable, less
</I>&gt;<i> compatible, and less reliable than simply using exceptions
</I>&gt;<i> integrated with the existing POSIX cleanup mechanism.
</I>
That's a rather broad sweep of the judgement brush, accompanied by
little to lend any insight into why you think it's so.

&gt;<i> Jason's model, putting together the puzzle pieces from various
</I>&gt;<i> messages here, appears to be basing cancel on an ordinary
</I>&gt;<i> (&quot;finalizable&quot;) C++ exception, which would be raised by any and all
</I>&gt;<i> cancellation points; but if the exception object were to be destroyed
</I>&gt;<i> (by finalization), the &quot;pending cancel&quot; would (might?) be
</I>&gt;<i> reasserted. I personally feel that it ought to be reasonable to
</I>&gt;<i> finalize the cancel and continue operation. If that wasn't really what
</I>&gt;<i> the application intended then it's an application error; but there are
</I>&gt;<i> cases where it's arguably reasonable, and the capability falls
</I>&gt;<i> naturally out of the model. 
</I>
I don't have a strong opinion about that, but I'm sympathetic to your
POV.

&gt;<i> Besides, &quot;reasserting cancel&quot; in all the right places and no other
</I>&gt;<i> places sounds like one of those projects that ends up being a lot
</I>&gt;<i> more subtle and error-prone than anyone ever expects. ;-)
</I>
I don't really see what you think is complicated about this.  In
Jason's model cancellation gets reasserted as soon as a caught
cancellation exception is destroyed by falling off the end of the
catch block without rethrowing.

&gt;&gt;&gt;<i>But I was really responding to the ideas about turning cancel into a
</I>&gt;&gt;&gt;<i>return status
</I>&gt;&gt;&gt;<i>
</I>&gt;&gt;&gt;<i>
</I>&gt;&gt;<i>...only for 'C' functions which ordinarily have a return status and don't
</I>&gt;&gt;<i>throw...  did you miss that?
</I>&gt;&gt;<i>
</I>&gt;<i> It doesn't matter. There are many problems lurking here, and only one
</I>&gt;<i> of the more obvious is that this attempt to avoid breaking C code that
</I>&gt;<i> depends on 'if (error) {cleanup(); return error;}' will still break if
</I>&gt;<i> 'cleanup()' depends on any cancellation points... which in fact is
</I>&gt;<i> quite likely if 'error' originates from a cancellation point (e.g.,
</I>&gt;<i> I/O). 
</I>
In what sense will it break?

&gt;<i> I also don't believe that a sufficient body of library code
</I>&gt;<i> really follows the simple and direct 'if (error) return error;'
</I>&gt;<i> pattern to make this strategy particularly useful even without the
</I>&gt;<i> cleanup issues -- but that's a different matter. 
</I>
Before I used exceptions I certainly wrote reams and reams of code
just like that.

&gt;<i> (It might actually be
</I>&gt;<i> better if the pending cancel wasn't &quot;sticky&quot;, 
</I>
You lost me.  What's &quot;sticky&quot;?

&gt;<i> since then the cleanup at least might succeed; but then you'll lose
</I>&gt;<i> cancels all over the place, and that's definitely not
</I>&gt;<i> acceptable. 
</I>
AFAICT you haven't understood Jason's proposal at all, though I may be
misinterpreting your response.

&gt;<i> It's fine for a thread to CHOOSE not to accept a cancel,
</I>&gt;<i> either by keeping cancellation disabled or by not calling a
</I>&gt;<i> cancellation point; but it's not OK to lose a delivered cancel. You
</I>&gt;<i> can declare this to be the application/library developer's job
</I>&gt;<i> rather than the implementation's job -- but that just means all this
</I>&gt;<i> &quot;broken&quot; C code that fails to propagate errors needs to be
</I>&gt;<i> redesigned; and you haven't solved any real problem. That's the
</I>&gt;<i> basic point: it needs to be analyzed and likely redesigned to be
</I>&gt;<i> cancel-safe. You simply can't avoid that requirement.)
</I>&gt;<i>
</I>&gt;<i> A) You cannot transparently make non cancel-safe C code cancel-safe,
</I>&gt;<i> no matter how you try.
</I>
It depends how &quot;cancel&quot; is implemented I guess.

&gt;<i> B) Delivering cancellation as a C++ exception will make much (though
</I>&gt;<i> not all) exception-safe C++ &quot;translucently&quot; cancel-safe.
</I>
Sure, provided that you only deliver it from functions which could
otherwise throw.

&gt;<i> C) POSIX C cancel-safe code is already cancel-safe, but for an
</I>&gt;<i> application containing mixed call stacks to remain cancel-safe, POSIX
</I>&gt;<i> cleanup and C++ exceptions must be integrated into a single unwind;
</I>&gt;<i> i.e., (by whatever name) basing both on a common exception library.
</I>&gt;<i> D) There are no useful or interestingly large bodies of rigorously
</I>&gt;<i> &quot;cancel-safe&quot; C code that are not POSIX cancel-safe.
</I>&gt;<i>
</I>&gt;<i> In other words, cancel-as-exception, and pthread_cleanup_push as a
</I>&gt;<i> C-based &quot;try&quot; extension, gives you all the cancel-safety you can
</I>&gt;<i> reasonably expect without recoding (and likely redesign). That's all
</I>&gt;<i> that's needed. No less will suffice, and any more is needless
</I>&gt;<i> complication. 
</I>
Except for the fact that we have normally-non-throwing functions that
have started to throw, I don't think anyone's disagreeing with you.

-- 
Dave Abrahams
Boost Consulting
www.boost-consulting.com


</PRE>

<!--endarticle-->
    <HR>
    <P><UL>
        <!--threads-->
	<LI>Previous message: <A HREF="000250.html">[c++-pthreads] Re: thread-safety definition
</A></li>
	<LI>Next message: <A HREF="000187.html">[c++-pthreads] Re: thread-safety definition
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#186">[ date ]</a>
              <a href="thread.html#186">[ thread ]</a>
              <a href="subject.html#186">[ subject ]</a>
              <a href="author.html#186">[ author ]</a>
         </LI>
       </UL>

<hr>
<a href="http://sourcerytools.com/cgi-bin/mailman/listinfo/c++-pthreads">More information about the c++-pthreads
mailing list</a><br>
</body></html>
